File: aihound/scanners/openai_cli.py
"""Scanner for OpenAI / Codex CLI credentials."""

from __future__ import annotations

import json
import logging
from pathlib import Path
from typing import Optional

logger = logging.getLogger("aihound.scanners.openai_cli")

from aihound.core.scanner import (
    BaseScanner, CredentialFinding, ScanResult, StorageType,
)
from aihound.core.platform import (
    detect_platform, Platform, get_home, get_appdata, get_wsl_windows_home,
)
from aihound.core.redactor import mask_value
from aihound.core.permissions import (
    get_file_permissions, get_file_owner, assess_risk,
    get_file_mtime, describe_staleness,
)
from aihound.remediation import hint_migrate_to_env
from aihound.scanners import register


SECRET_KEY_TOKENS = ("token", "key", "secret", "api_key", "apikey", "access_key", "refresh")


@register
class OpenAICLIScanner(BaseScanner):
    def name(self) -> str:
        return "OpenAI/Codex CLI"

    def slug(self) -> str:
        return "openai-cli"

    def scan(self, show_secrets: bool = False) -> ScanResult:
        plat = detect_platform()
        result = ScanResult(scanner_name=self.name(), platform=plat.value)

        # Plaintext api_key files
        for path in self._get_plaintext_paths(plat):
            self._scan_plaintext(path, result, show_secrets)

        # JSON config files (single files)
        for path in self._get_json_paths(plat):
            self._scan_json_file(path, result, show_secrets)

        # Directory glob: .codex dirs and %APPDATA%/OpenAI dirs
        for directory in self._get_json_directories(plat):
            if not directory.exists() or not directory.is_dir():
                continue
            try:
                for json_path in directory.glob("*.json"):
                    self._scan_json_file(json_path, result, show_secrets)
            except OSError as e:
                result.errors.append(f"Failed to enumerate {directory}: {e}")

        return result

    def _get_plaintext_paths(self, plat: Platform) -> list[Path]:
        paths: list[Path] = []
        home = get_home()

        paths.append(home / ".openai" / "api_key")

        if plat == Platform.WSL:
            win_home = get_wsl_windows_home()
            if win_home:
                paths.append(win_home / ".openai" / "api_key")
            appdata = get_appdata()
            if appdata:
                paths.append(appdata / "OpenAI" / "api_key")
        elif plat == Platform.WINDOWS:
            appdata = get_appdata()
            if appdata:
                paths.append(appdata / "OpenAI" / "api_key")

        return paths

    def _get_json_paths(self, plat: Platform) -> list[Path]:
        paths: list[Path] = []
        home = get_home()

        paths.append(home / ".openai" / "auth.json")

        if plat == Platform.WSL:
            win_home = get_wsl_windows_home()
            if win_home:
                paths.append(win_home / ".openai" / "auth.json")
            appdata = get_appdata()
            if appdata:
                paths.append(appdata / "OpenAI" / "auth.json")
        elif plat == Platform.WINDOWS:
            appdata = get_appdata()
            if appdata:
                paths.append(appdata / "OpenAI" / "auth.json")

        return paths

    def _get_json_directories(self, plat: Platform) -> list[Path]:
        dirs: list[Path] = []
        home = get_home()

        dirs.append(home / ".codex")

        if plat == Platform.WSL:
            win_home = get_wsl_windows_home()
            if win_home:
                dirs.append(win_home / ".codex")
            appdata = get_appdata()
            if appdata:
                dirs.append(appdata / "OpenAI")
        elif plat == Platform.WINDOWS:
            appdata = get_appdata()
            if appdata:
                dirs.append(appdata / "OpenAI")

        return dirs

    def _scan_plaintext(
        self, path: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        if not path.exists():
            logger.debug("Plaintext key file not found: %s", path)
            return

        logger.debug("Reading plaintext key file: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)
        storage = StorageType.PLAINTEXT_FILE

        try:
            value = path.read_text(encoding="utf-8").strip()
        except OSError as e:
            logger.warning("Failed to read %s: %s", path, e)
            result.errors.append(f"Failed to read {path}: {e}")
            return

        if not value:
            return

        notes: list[str] = []
        if mtime:
            notes.append(f"File last modified: {describe_staleness(mtime)}")

        result.findings.append(CredentialFinding(
            tool_name=self.name(),
            credential_type="openai_api_key",
            storage_type=storage,
            location=str(path),
            exists=True,
            risk_level=assess_risk(storage, path),
            value_preview=mask_value(value, show_full=show_secrets),
            raw_value=value if show_secrets else None,
            file_permissions=perms,
            file_owner=owner,
            file_modified=mtime,
            remediation="Use OPENAI_API_KEY environment variable instead of plaintext file",
            remediation_hint=hint_migrate_to_env(["OPENAI_API_KEY"], str(path)),
            notes=notes,
        ))

    def _scan_json_file(
        self, path: Path, result: ScanResult, show_secrets: bool
    ) -> None:
        if not path.exists():
            logger.debug("OpenAI JSON file not found: %s", path)
            return

        logger.debug("Reading OpenAI JSON file: %s", path)
        perms = get_file_permissions(path)
        owner = get_file_owner(path)
        mtime = get_file_mtime(path)

        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError) as e:
            logger.warning("Failed to parse %s: %s", path, e)
            result.errors.append(f"Failed to parse {path}: {e}")
            return

        self._walk_json(data, path, perms, owner, mtime, result, show_secrets, key_path="")

    def _walk_json(
        self,
        data,
        path: Path,
        perms: Optional[str],
        owner: Optional[str],
        mtime,
        result: ScanResult,
        show_secrets: bool,
        key_path: str,
    ) -> None:
        storage = StorageType.PLAINTEXT_JSON

        if isinstance(data, dict):
            for k, v in data.items():
                if not isinstance(k, str):
                    continue
                sub_path = f"{key_path}.{k}" if key_path else k

                if isinstance(v, str) and v:
                    k_lower = k.lower()
                    if any(tok in k_lower for tok in SECRET_KEY_TOKENS) and len(v) > 8:
                        notes = [f"JSON key path: {sub_path}"]
                        if mtime:
                            notes.append(f"File last modified: {describe_staleness(mtime)}")
                        result.findings.append(CredentialFinding(
                            tool_name=self.name(),
                            credential_type=f"openai:{k}",
                            storage_type=storage,
                            location=str(path),
                            exists=True,
                            risk_level=assess_risk(storage, path),
                            value_preview=mask_value(v, show_full=show_secrets),
                            raw_value=v if show_secrets else None,
                            file_permissions=perms,
                            file_owner=owner,
                            file_modified=mtime,
                            remediation="Use OPENAI_API_KEY environment variable instead of plaintext file",
                            remediation_hint=hint_migrate_to_env(["OPENAI_API_KEY"], str(path)),
                            notes=notes,
                        ))
                elif isinstance(v, (dict, list)):
                    self._walk_json(v, path, perms, owner, mtime, result, show_secrets, sub_path)
        elif isinstance(data, list):
            for i, item in enumerate(data):
                sub_path = f"{key_path}[{i}]"
                if isinstance(item, (dict, list)):
                    self._walk_json(item, path, perms, owner, mtime, result, show_secrets, sub_path)

File: Other Versions/Go/scanners/openai_cli.go
package scanners

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"aihound/core"
	"aihound/remediation"
)

var openaiSecretKeyTokens = []string{"token", "key", "secret", "api_key", "apikey", "access_key", "refresh"}

type openaiCLIScanner struct{}

func init() {
	Register(&openaiCLIScanner{})
}

func (s *openaiCLIScanner) Name() string        { return "OpenAI/Codex CLI" }
func (s *openaiCLIScanner) Slug() string         { return "openai-cli" }
func (s *openaiCLIScanner) IsApplicable() bool   { return true }

func (s *openaiCLIScanner) Scan(showSecrets bool) core.ScanResult {
	result := core.ScanResult{ScannerName: s.Name(), Platform: core.DetectPlatform().String()}
	plat := core.DetectPlatform()

	for _, p := range s.getPlaintextPaths(plat) {
		s.scanPlaintext(p, &result, showSecrets)
	}

	for _, p := range s.getJSONPaths(plat) {
		s.scanJSONFile(p, &result, showSecrets)
	}

	for _, dir := range s.getJSONDirectories(plat) {
		info, err := os.Stat(dir)
		if err != nil || !info.IsDir() {
			continue
		}
		matches, err := filepath.Glob(filepath.Join(dir, "*.json"))
		if err != nil {
			result.Errors = append(result.Errors, fmt.Sprintf("Failed to enumerate %s: %v", dir, err))
			continue
		}
		for _, jsonFile := range matches {
			s.scanJSONFile(jsonFile, &result, showSecrets)
		}
	}

	return result
}

func (s *openaiCLIScanner) getPlaintextPaths(plat core.Platform) []string {
	var paths []string
	home := core.GetHome()
	paths = append(paths, filepath.Join(home, ".openai", "api_key"))

	if plat == core.PlatformWSL {
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			paths = append(paths, filepath.Join(winHome, ".openai", "api_key"))
		}
		if appdata := core.GetAppData(); appdata != "" {
			paths = append(paths, filepath.Join(appdata, "OpenAI", "api_key"))
		}
	} else if plat == core.PlatformWindows {
		if appdata := core.GetAppData(); appdata != "" {
			paths = append(paths, filepath.Join(appdata, "OpenAI", "api_key"))
		}
	}
	return paths
}

func (s *openaiCLIScanner) getJSONPaths(plat core.Platform) []string {
	var paths []string
	home := core.GetHome()
	paths = append(paths, filepath.Join(home, ".openai", "auth.json"))

	if plat == core.PlatformWSL {
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			paths = append(paths, filepath.Join(winHome, ".openai", "auth.json"))
		}
		if appdata := core.GetAppData(); appdata != "" {
			paths = append(paths, filepath.Join(appdata, "OpenAI", "auth.json"))
		}
	} else if plat == core.PlatformWindows {
		if appdata := core.GetAppData(); appdata != "" {
			paths = append(paths, filepath.Join(appdata, "OpenAI", "auth.json"))
		}
	}
	return paths
}

func (s *openaiCLIScanner) getJSONDirectories(plat core.Platform) []string {
	var dirs []string
	home := core.GetHome()
	dirs = append(dirs, filepath.Join(home, ".codex"))

	if plat == core.PlatformWSL {
		if winHome := core.GetWSLWindowsHome(); winHome != "" {
			dirs = append(dirs, filepath.Join(winHome, ".codex"))
		}
		if appdata := core.GetAppData(); appdata != "" {
			dirs = append(dirs, filepath.Join(appdata, "OpenAI"))
		}
	} else if plat == core.PlatformWindows {
		if appdata := core.GetAppData(); appdata != "" {
			dirs = append(dirs, filepath.Join(appdata, "OpenAI"))
		}
	}
	return dirs
}

func (s *openaiCLIScanner) scanPlaintext(path string, result *core.ScanResult, showSecrets bool) {
	data, err := os.ReadFile(path)
	if err != nil {
		return
	}

	value := strings.TrimSpace(string(data))
	if value == "" {
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)
	storage := core.PlaintextFile

	rawValue := ""
	if showSecrets {
		rawValue = value
	}

	var notes []string
	if mtime := core.GetFileMtimeTime(path); !mtime.IsZero() {
		notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
	}

	result.Findings = append(result.Findings, core.CredentialFinding{
		ToolName:        s.Name(),
		CredentialType:  "openai_api_key",
		StorageType:     storage,
		Location:        path,
		Exists:          true,
		RiskLevel:       core.AssessRisk(storage, path),
		ValuePreview:    core.MaskValue(value, showSecrets),
		RawValue:        rawValue,
		FilePermissions: perms,
		FileOwner:       owner,
		FileModified:    core.GetFileMtime(path),
		Remediation:     "Use OPENAI_API_KEY environment variable instead of plaintext file",
		RemediationHint: remediation.HintMigrateToEnv([]string{"OPENAI_API_KEY"}, path),
		Notes:           notes,
	})
}

func (s *openaiCLIScanner) scanJSONFile(path string, result *core.ScanResult, showSecrets bool) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return
	}

	var data interface{}
	if err := json.Unmarshal(raw, &data); err != nil {
		result.Errors = append(result.Errors, fmt.Sprintf("Failed to parse %s: %v", path, err))
		return
	}

	perms := core.GetFilePermissions(path)
	owner := core.GetFileOwner(path)
	s.walkJSON(data, path, perms, owner, result, showSecrets, "")
}

func (s *openaiCLIScanner) walkJSON(data interface{}, path, perms, owner string, result *core.ScanResult, showSecrets bool, keyPath string) {
	storage := core.PlaintextJSON

	switch v := data.(type) {
	case map[string]interface{}:
		for k, val := range v {
			subPath := k
			if keyPath != "" {
				subPath = keyPath + "." + k
			}

			switch inner := val.(type) {
			case string:
				if inner == "" {
					continue
				}
				kLower := strings.ToLower(k)
				matched := false
				for _, tok := range openaiSecretKeyTokens {
					if strings.Contains(kLower, tok) {
						matched = true
						break
					}
				}
				if !matched || len(inner) <= 8 {
					continue
				}

				notes := []string{fmt.Sprintf("JSON key path: %s", subPath)}
				if mtime := core.GetFileMtimeTime(path); !mtime.IsZero() {
					notes = append(notes, "File last modified: "+core.DescribeStaleness(mtime))
				}

				rawValue := ""
				if showSecrets {
					rawValue = inner
				}

				result.Findings = append(result.Findings, core.CredentialFinding{
					ToolName:        s.Name(),
					CredentialType:  fmt.Sprintf("openai:%s", k),
					StorageType:     storage,
					Location:        path,
					Exists:          true,
					RiskLevel:       core.AssessRisk(storage, path),
					ValuePreview:    core.MaskValue(inner, showSecrets),
					RawValue:        rawValue,
					FilePermissions: perms,
					FileOwner:       owner,
					FileModified:    core.GetFileMtime(path),
					Remediation:     "Use OPENAI_API_KEY environment variable instead of plaintext file",
					RemediationHint: remediation.HintMigrateToEnv([]string{"OPENAI_API_KEY"}, path),
					Notes:           notes,
				})
			case map[string]interface{}, []interface{}:
				s.walkJSON(inner, path, perms, owner, result, showSecrets, subPath)
			}
		}
	case []interface{}:
		for i, item := range v {
			subPath := fmt.Sprintf("%s[%d]", keyPath, i)
			switch item.(type) {
			case map[string]interface{}, []interface{}:
				s.walkJSON(item, path, perms, owner, result, showSecrets, subPath)
			}
		}
	}
}
